### PR TITLE
[mlir][affine] remove divide zero check when simplifer affineMap (#64622)

### PR DESCRIPTION
**Pull request by lipracer**
When performing constant folding on the affineApplyOp, there is a division of 0 in the affine map.
[related issue](https://github.com/llvm/llvm-project/issues/64622)

### CODE REVIEW

**Comment by lipracer** [comment]
TODO:
1. add a test
2. `SimpleAffineExprFlattener` this class still has crashes
I'm not sure if this is a reasonable solution.

**Comment by matthias-springer** [comment]
I have not looked at the fix in detail yet, just one comment: This bug looks like a case of "folding leads to invalid op". But we do not create an invalid op but already crash when trying to create it. Similar issues were discussed in the context of "poison semantics". You could have IR such as:

```mlir
scf.if %c {
  // The folder of affine.apply crashes due to a division by 0, but %c is such that the branch would never be executed if there would be a division by zero. E.g., %c could be the result of a size check.
  affine.apply
}
```

I think the consensus on such issues was, if a folder that turns dynamic information into static information (I think that's what this folder here does) would create an invalid op, then the op should not be folded. Same in this case, we should not fold the op, but also not crash. We probably also don't want to emit warnings.

**Comment by lipracer** [comment]
I am found the poison-dem https://mlir.llvm.org/OpenMeetings/2022-11-03-Poison-Semantics-For-MLIR.pdf, just remove runtime check,maybe we need define freeze and poison value for affine map?

**Review by joker-eph — COMMENTED** [review]
Can you please update the PR title to be descriptive? See https://mlir.llvm.org/getting_started/Contributing/#commit-messages

**Comment by lipracer** [comment]
When `AffineApplyOp::fold` return `PoisonAttr`, the test case crash and tell me UBDialect need to initilize.

**Comment by matthias-springer** [comment]
> When `AffineApplyOp::fold` return `PoisonAttr`, the test case crash and tell me UBDialect need to initilize.

That means the UB dialect would have to be declared as a dependent dialect of the Affine dialect (because such ops are created by the canonicalization). I am not an expert on poison semantics, I'd ask on Discord to see if that's desirable and the recommended path. Also, @kuhar may have an opinion on this.

**Comment by lipracer** [comment]
Perhaps the arith dialect can fold out 'poison' semantics, is this pull request ready to be merged?

File: mlir/lib/IR/AffineExpr.cpp [hunk 1]
```diff
@@ -1326,14 +1334,13 @@ void SimpleAffineExprFlattener::visitDivExpr(AffineBinaryOpExpr expr,
                                              localExprs, context);
```
Comment (matthias-springer) [inline mlir/lib/IR/AffineExpr.cpp:1334]: Can we also go into this branch if `getRHS` is a constant expr that is `0`?

**Review by matthias-springer — COMMENTED** [review]
I'm a bit hesitant to add a `LogicalResult` return value to the entire visitor logic here. It is not clear to me whether this operation should fold to a poison value in case of a division by zero. If that's not the case, we won't need the `LogicalResult` return value. Until that's clear, I would just go with a simpler fix: just stop folding when there's a division by zero, by treating it the same was as "not a constant".

**Comment by matthias-springer** [comment]
That being said, don't throw this code away yet, if the consensus is that folding/canonicalization should create poison values, we probably need exactly the code that you wrote here.

**Comment by lipracer** [comment]
> I'm a bit hesitant to add a `LogicalResult` return value to the entire visitor logic here. It is not clear to me whether this operation should fold to a poison value in case of a division by zero. If that's not the case, we won't need the `LogicalResult` return value. Until that's clear, I would just go with a simpler fix: just stop folding when there's a division by zero, by treating it the same was as "not a constant".
Perhaps `mlir::WalkResult` is more suitable here, but it seems to be used to traverse ir.

File: mlir/lib/IR/AffineExpr.cpp (hunk 1 above)
Comment (lipracer) [reply mlir/lib/IR/AffineExpr.cpp:1334]: run [issue](https://github.com/llvm/llvm-project/issues/64622) test case, will going into this branch.

**Comment by lipracer** [comment]
I have found a fix for not being able to create ‘poisonstr’. ‘ubdialect’ is lazy to load and seems to only occur during the parse phase, so I can load it during the initialization phase.

**Comment by lipracer** [comment]
@joker-eph Can you help me review it?

File: mlir/include/mlir/IR/AffineExprVisitor.h [hunk 2]
```diff
@@ -65,8 +66,80 @@ namespace mlir {
...
+class AffineExprVisitorBase {
+public:
+  // Function to visit an AffineExpr.
+  RetTy visit(AffineExpr expr) {
+    static_assert(std::is_base_of<AffineExprVisitorBase, SubClass>::value,
+                  "Must instantiate with a derived type of AffineExprVisitor");
+    switch (expr.getKind()) {
+    case AffineExprKind::Add: {
+      auto binOpExpr = expr.cast<AffineBinaryOpExpr>();
+      return static_cast<SubClass *>(this)->visitAddExpr(binOpExpr);
```
Comment (kuhar) [inline mlir/include/mlir/IR/AffineExprVisitor.h:79]: nit: consider hoisting `static_cast<SubClass *>(this)` to a local variable, since it repeats in every single `case`  body

File: mlir/lib/Analysis/FlatLinearValueConstraints.cpp [hunk 3]
```diff
@@ -86,7 +86,8 @@ getFlattenedAffineExprs(ArrayRef<AffineExpr> exprs, unsigned numDims,
     if (!expr.isPureAffine())
       return failure();
 
-    flattener.walkPostOrder(expr);
+    auto flattenResult = flattener.walkPostOrder(expr);
+    assert(succeeded(flattenResult) && "affine expr containts poison expr");
```
Comment (kuhar) [inline mlir/lib/Analysis/FlatLinearValueConstraints.cpp:90]: this will cause unused variable warnings in non-debug builds
```suggestion
    (void)flattenResult;
    assert(succeeded(flattenResult) && "affine expr containts poison expr");
```

Comment (kuhar) [inline mlir/lib/Analysis/FlatLinearValueConstraints.cpp:90]: Why does this warrant an assertion? Could we choose to materialize a different constant?

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 4]
```diff
@@ -700,6 +708,94 @@ static std::optional<int64_t> getUpperBound(Value iv) {
...
+        auto ub = getBoundForExpr(binOpExpr.getLHS(), numDims, numSymbols,
+                                  constLowerBounds, constUpperBounds, isUpper);
+        if (ub && lb &&
+            floorDiv(*lb, rhsConstVal) == floorDiv(*ub, rhsConstVal))
+          return isUpper ? mod(*ub, rhsConstVal) : mod(*lb, rhsConstVal);
+        return isUpper ? rhsConstVal - 1 : 0;
+      }
+    }
+  }
+  // Flatten the expression.
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:769]:
```suggestion

  // Flatten the expression.
```

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 5]
```diff
@@ -700,6 +708,94 @@ static std::optional<int64_t> getUpperBound(Value iv) {
...
+  SimpleAffineExprFlattener flattener(numDims, numSymbols);
+  auto flattenResult = flattener.walkPostOrder(expr);
+  assert(succeeded(flattenResult) && "affine expr containts poison expr");
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:772]: same here

File: mlir/lib/IR/AffineMap.cpp [hunk 6]
```diff
@@ -57,13 +60,34 @@ class AffineExprConstantFolder {
...
+              return std::nullopt;
+            }
+            return mod(lhs, rhs);
+          });
     case AffineExprKind::FloorDiv:
       return constantFoldBinExpr(
-          expr, [](int64_t lhs, int64_t rhs) { return floorDiv(lhs, rhs); });
+          expr,
+          [expr, this](int64_t lhs, int64_t rhs) -> std::optional<int64_t> {
+            if (0 == rhs) {
```
Comment (kuhar) [inline mlir/lib/IR/AffineMap.cpp:75]:
```suggestion
            if (rhs == 0) {
```

File: mlir/lib/IR/AffineMap.cpp [hunk 7]
```diff
@@ -57,13 +60,34 @@ class AffineExprConstantFolder {
...
+              return std::nullopt;
+            }
+            return floorDiv(lhs, rhs);
+          });
     case AffineExprKind::CeilDiv:
       return constantFoldBinExpr(
-          expr, [](int64_t lhs, int64_t rhs) { return ceilDiv(lhs, rhs); });
+          expr,
+          [expr, this](int64_t lhs, int64_t rhs) -> std::optional<int64_t> {
+            if (0 == rhs) {
```
Comment (kuhar) [inline mlir/lib/IR/AffineMap.cpp:85]:
```suggestion
            if (rhs == 0) {
```

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 8]
```diff
@@ -700,6 +708,94 @@ static std::optional<int64_t> getUpperBound(Value iv) {
...
+                bool isUpper) {
+  // Handle divs and mods.
+  if (auto binOpExpr = expr.dyn_cast<AffineBinaryOpExpr>()) {
+    // If the LHS of a floor or ceil is bounded and the RHS is a constant, we
+    // can compute an upper bound.
+    if (binOpExpr.getKind() == AffineExprKind::FloorDiv) {
+      auto rhsConst = binOpExpr.getRHS().dyn_cast<AffineConstantExpr>();
+      if (!rhsConst || rhsConst.getValue() < 1)
+        return std::nullopt;
+      auto bound = getBoundForExpr(binOpExpr.getLHS(), numDims, numSymbols,
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:732]: Could you write the full type here? It's not obvious just based on the RHS.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 9]
```diff
@@ -700,6 +708,94 @@ static std::optional<int64_t> getUpperBound(Value iv) {
...
+                                   constLowerBounds, constUpperBounds, isUpper);
+      if (!bound)
+        return std::nullopt;
+      return mlir::floorDiv(*bound, rhsConst.getValue());
+    }
+    if (binOpExpr.getKind() == AffineExprKind::CeilDiv) {
+      auto rhsConst = binOpExpr.getRHS().dyn_cast<AffineConstantExpr>();
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:739]: use `dyn_cast<T>(expr)`: https://mlir.llvm.org/deprecation/

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 10]
```diff
@@ -700,6 +708,94 @@ static std::optional<int64_t> getUpperBound(Value iv) {
...
+      if (rhsConst && rhsConst.getValue() >= 1) {
+        auto bound =
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:741]: also here

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 11]
```diff
@@ -700,6 +708,94 @@ static std::optional<int64_t> getUpperBound(Value iv) {
...
+      // lhs mod c is always <= c - 1 and non-negative. In addition, if `lhs` is
+      // bounded such that lb <= lhs <= ub and lb floordiv c == ub floordiv c
+      // (same "interval"), then lb mod c <= lhs mod c <= ub mod c.
+      auto rhsConst = binOpExpr.getRHS().dyn_cast<AffineConstantExpr>();
+      if (rhsConst && rhsConst.getValue() >= 1) {
+        int64_t rhsConstVal = rhsConst.getValue();
+        auto lb = getBoundForExpr(binOpExpr.getLHS(), numDims, numSymbols,
+                                  constLowerBounds, constUpperBounds,
+                                  /*isUpper=*/false);
+        auto ub = getBoundForExpr(binOpExpr.getLHS(), numDims, numSymbols,
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:760]: also here

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 12]
```diff
@@ -700,6 +708,94 @@ static std::optional<int64_t> getUpperBound(Value iv) {
...
+        if (ub && lb &&
+            floorDiv(*lb, rhsConstVal) == floorDiv(*ub, rhsConstVal))
+          return isUpper ? mod(*ub, rhsConstVal) : mod(*lb, rhsConstVal);
+        return isUpper ? rhsConstVal - 1 : 0;
+      }
+    }
+  }
+  // Flatten the expression.
+  SimpleAffineExprFlattener flattener(numDims, numSymbols);
+  auto flattenResult = flattener.walkPostOrder(expr);
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:771]: also here

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 13]
```diff
@@ -3379,7 +3475,8 @@ static LogicalResult canonicalizeMapExprAndTermOrder(AffineMap &map) {
       return failure();
 
     SimpleAffineExprFlattener flattener(map.getNumDims(), map.getNumSymbols());
-    flattener.walkPostOrder(resultExpr);
+    auto flattenResult = flattener.walkPostOrder(resultExpr);
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:3478]: also here

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 14]
```diff
@@ -580,7 +583,12 @@ OpFoldResult AffineApplyOp::fold(FoldAdaptor adaptor) {
 
   // Otherwise, default to folding the map.
   SmallVector<Attribute, 1> result;
-  if (failed(map.constantFold(adaptor.getMapOperands(), result)))
+  bool hasPoison = false;
+  auto foldResult =
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:587]: also here

File: mlir/test/Dialect/Affine/constant-fold.mlir [hunk 15]
```diff
@@ -60,3 +60,22 @@ func.func @affine_min(%variable: index) -> (index, index) {
...
   return %0, %1 : index, index
 }
+
+// -----
+
+func.func @affine_apply_poison_division_zero() {
+  %c0 = arith.constant 0 : index
+  %0 = affine.apply affine_map<(d0)[s0] -> (d0 mod s0)>(%c0)[%c0]
+  %1 = affine.apply affine_map<(d0)[s0] -> (d0 floordiv s0)>(%c0)[%c0]
+  %2 = affine.apply affine_map<(d0)[s0] -> (d0 ceildiv s0)>(%c0)[%c0]
```
Comment (kuhar) [inline mlir/test/Dialect/Affine/constant-fold.mlir:70]: This makes sense to me, the symbol could come from some other simplification and it would be better not to trigger a verifier failure

**Review by kuhar — CHANGES_REQUESTED** [review]

File: mlir/lib/IR/AffineMap.cpp (hunk 6 above)
Comment (lipracer) [reply mlir/lib/IR/AffineMap.cpp:75]: Is it more reasonable to place constants on the left?

Comment (lipracer) [reply mlir/lib/IR/AffineMap.cpp:75]: > 

Is it more reasonable to place constants on the left?

Comment (kuhar) [reply mlir/lib/IR/AffineMap.cpp]: It looks unusual. Why would you do that? If that's to guard against accidental assignment, the warnings enabled by default in llvm already enforce that assignments are wrapped with an extra layer of parens.

Comment (lipracer) [reply mlir/lib/IR/AffineMap.cpp]: Thanks.

**Comment by lipracer** [comment]
There is an additional discussion on whether defining a ‘poison expression’ would be more reasonable, as it would avoid many simple checks?

**Comment by lipracer** [comment]
I will submit other patch for cleanup the member function with T.cast style.

File: mlir/include/mlir/IR/AffineExprVisitor.h [hunk 16]
```diff
@@ -75,117 +146,111 @@ class AffineExprVisitor {
...
     case AffineExprKind::CeilDiv: {
       auto binOpExpr = expr.cast<AffineBinaryOpExpr>();
       walkOperandsPostOrder(binOpExpr);
-      return static_cast<SubClass *>(this)->visitCeilDivExpr(binOpExpr);
+      return self->visitCeilDivExpr(binOpExpr);
     }
     case AffineExprKind::Constant:
-      return static_cast<SubClass *>(this)->visitConstantExpr(
-          expr.cast<AffineConstantExpr>());
+      return self->visitConstantExpr(expr.cast<AffineConstantExpr>());
```
Comment (jsetoain) [inline mlir/include/mlir/IR/AffineExprVisitor.h:177]: Using `cast` method [instead of the free function](https://mlir.llvm.org/deprecation/).

File: mlir/include/mlir/IR/AffineExprVisitor.h [hunk 17]
```diff
@@ -75,117 +146,111 @@ class AffineExprVisitor {
...
     case AffineExprKind::DimId:
-      return static_cast<SubClass *>(this)->visitDimExpr(
-          expr.cast<AffineDimExpr>());
+      return self->visitDimExpr(expr.cast<AffineDimExpr>());
```
Comment (jsetoain) [inline mlir/include/mlir/IR/AffineExprVisitor.h:179]: [Use the free function instead of the cast method](https://mlir.llvm.org/deprecation/).

File: mlir/include/mlir/IR/AffineExprVisitor.h [hunk 18]
```diff
@@ -75,117 +146,111 @@ class AffineExprVisitor {
...
     case AffineExprKind::SymbolId:
-      return static_cast<SubClass *>(this)->visitSymbolExpr(
-          expr.cast<AffineSymbolExpr>());
+      return self->visitSymbolExpr(expr.cast<AffineSymbolExpr>());
```
Comment (jsetoain) [inline mlir/include/mlir/IR/AffineExprVisitor.h:181]: [Use the free function instead of the cast method](https://mlir.llvm.org/deprecation/).

File: mlir/include/mlir/IR/AffineExprVisitor.h [hunk 19]
```diff
@@ -75,117 +146,111 @@ class AffineExprVisitor {
...
       auto binOpExpr = expr.cast<AffineBinaryOpExpr>();
-      return static_cast<SubClass *>(this)->visitCeilDivExpr(binOpExpr);
+      if (failed(walkOperandsPostOrder(binOpExpr)))
+        return failure();
+      return self->visitCeilDivExpr(binOpExpr);
     }
     case AffineExprKind::Constant:
-      return static_cast<SubClass *>(this)->visitConstantExpr(
-          expr.cast<AffineConstantExpr>());
+      return self->visitConstantExpr(expr.cast<AffineConstantExpr>());
```
Comment (jsetoain) [inline mlir/include/mlir/IR/AffineExprVisitor.h:237]: [Use the free function instead of the cast method](https://mlir.llvm.org/deprecation/).

File: mlir/include/mlir/IR/AffineExprVisitor.h [hunk 20]
```diff
@@ -75,117 +146,111 @@ class AffineExprVisitor {
...
     case AffineExprKind::DimId:
-      return static_cast<SubClass *>(this)->visitDimExpr(
-          expr.cast<AffineDimExpr>());
+      return self->visitDimExpr(expr.cast<AffineDimExpr>());
```
Comment (jsetoain) [inline mlir/include/mlir/IR/AffineExprVisitor.h:239]: [Use the free function instead of the cast method](https://mlir.llvm.org/deprecation/).

File: mlir/include/mlir/IR/AffineExprVisitor.h [hunk 21]
```diff
@@ -75,117 +146,111 @@ class AffineExprVisitor {
...
     case AffineExprKind::SymbolId:
-      return static_cast<SubClass *>(this)->visitSymbolExpr(
-          expr.cast<AffineSymbolExpr>());
+      return self->visitSymbolExpr(expr.cast<AffineSymbolExpr>());
```
Comment (jsetoain) [inline mlir/include/mlir/IR/AffineExprVisitor.h:241]: [Use the free function instead of the cast method](https://mlir.llvm.org/deprecation/).

**Review by jsetoain — CHANGES_REQUESTED** [review]

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 22]
```diff
@@ -700,6 +708,100 @@ static std::optional<int64_t> getUpperBound(Value iv) {
...
+  // TODO: Handle local variables. We can get hold of flattener.localExprs and
+  // get bound on the local expr recursively.
+  if (flattener.numLocals > 0)
+    return std::nullopt;
+  int64_t bound = 0;
+  // Substitute the constant lower or upper bound for the dimensional or
+  // symbolic input depending on `isUpper` to determine the bound.
+  for (unsigned i = 0, e = numDims + numSymbols; i < e; ++i) {
+    if (flattenedExpr[i] > 0) {
+      auto &constBound = isUpper ? constUpperBounds[i] : constLowerBounds[i];
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:789]: Use the full type if not obvious based on the RHS

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 23]
```diff
@@ -700,6 +708,100 @@ static std::optional<int64_t> getUpperBound(Value iv) {
...
+      if (!constBound)
+        return std::nullopt;
+      bound += *constBound * flattenedExpr[i];
+    } else if (flattenedExpr[i] < 0) {
+      auto &constBound = isUpper ? constLowerBounds[i] : constUpperBounds[i];
```
Comment (kuhar) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:794]: Use the full type if not obvious based on the RHS

File: mlir/lib/Analysis/FlatLinearValueConstraints.cpp [hunk 24]
```diff
@@ -85,8 +85,10 @@ getFlattenedAffineExprs(ArrayRef<AffineExpr> exprs, unsigned numDims,
   for (auto expr : exprs) {
     if (!expr.isPureAffine())
       return failure();
-
-    flattener.walkPostOrder(expr);
+    // has poison expression
+    auto flattenResult = flattener.walkPostOrder(expr);
+    if (failed(flattenResult))
+      return failure();
```
Comment (Groverkss) [inline mlir/lib/Analysis/FlatLinearValueConstraints.cpp:91]: This is a little weird... Failure on flattening usually means it's semi-affine.  Could you update the documentation for this function?

File: mlir/lib/Dialect/Affine/IR/CMakeLists.txt [hunk 25]
```diff
@@ -16,8 +16,9 @@ add_mlir_dialect_library(MLIRAffineDialect
   MLIRDialectUtils
   MLIRIR
   MLIRLoopLikeInterface
-  MLIRMemRefDialect
```
Comment (Groverkss) [inline mlir/lib/Dialect/Affine/IR/CMakeLists.txt:19]: Why was this moved? These should be in alphabetical order.

File: mlir/include/mlir/IR/AffineExprVisitor.h (hunk 16 above)
Comment (lipracer) [reply mlir/include/mlir/IR/AffineExprVisitor.h:177]: > Using `cast` method [instead of the free function](https://mlir.llvm.org/deprecation/).

If llvm:: cast is used, it will result in a compilation error: 'classof' is not a member of 'mlir:: AffineSymbolExpr'
Here we need to implement CastInfo<T, AffineExpr>, and I have submitted another [PR]( https://github.com/llvm/llvm-project/pull/71269)

**Review by jsetoain — APPROVED** [review]
Looks good to me. Wait for a couple more days in case the other reviewers have anything else to say, please. It is not a trivial change, and I may have overlooked something.
